proposals/cache_v2.md

```diff
@@ -0,0 +1,177 @@
+- The up-to-date check is tightly coupled to the implementation of the cache. This does not allow for easy switching of the strategy to use for up-to-date checks.
+- Retrieval of layers by layer entries currently involves checking through the entire metadata and matching against the layer entries stored in the metadata. This could become slow as the cache grows in size.
+- The current implementation tightly couples the actual storage implementation with the contents stored in the cache. This does not allow for easily implementing new storage engines with possibly different underlying storage systems.
+- Since the metadata is only written out at the end of a Jib execution (successful or failure), interrupted Jib executions would result in lost metadata updates.
+
+These problems should be resolved before [Jib as a library for Java](https://github.com/GoogleContainerTools/jib/issues/337) is released.
+
+## Proposal
+
+### Goals
+
+- Clearly-defined storage engine interface that decouples the query system from the storage engine.
+- A default storage engine implementation that stores layer data independently and can retrieve layers by the layer entries that built that layer in constant time.
+
+### Solution
+
+#### Storage engine interface
+
+There are 4 actions that would be performed against the storage engine:
+
+- Save a cache entry
+- List the entries in the cache
+- Retrieve a cache entry by layer digest
+- Retrieve a layer digest by a selector (explained below)
+- (Optional) Prune the cache
+
+Cache entry writes are provided with:
+
+- Layer blob (generates the layer digest, diff ID, and size)
+- Optional selector to additionally reference the layer
+- Optional metadata blob
+
+Cache entry reads are provided with:
+
+- Layer digest
+- Layer diff ID
+- Layer size
+- Layer blob
+- Optional metadata blob
+
+#### Storage engine implementation
+
+There are two types of layers - application layers and base image layers. Base image layers would only need to store their layer data (layer digest and diff ID). Application layers need to store their layer data along with metadata and a custom selector.
+
+In the provided storage implementation, the metadata will be just the last modified time of the layer (the latest modification time for the files that go into the layer). **The selector will be a digest of the layer entries that built the layer.**
+
+For example, the cache directory structure looks like:
+
+```
+layers/
+  36a2b7401dcddc50a35aeaa81085718b9d5fbce9d607c55a1d79beec2469f9ac/
+    326a609681777ee4ca02b1898579c9e07801ef066a629a3c59fa6df6ab42b7aa.layer
+    metadata
+selectors/
+  65de3b72aaf98e4f300ccdf7d64bf9a3b1e23c8c44a1242265f717db1a0877e9
+```
+
+The `layers/` directory consists of directories for each layer, named by the layer digest.
+Inside each of the layer directories:
+- The layer tarball is named by the layer diff ID with extension `.layer`
```

Comment:
Is this still the plan? Noticed you removed this extension here https://github.com/GoogleContainerTools/jib/pull/901/files#diff-68ffab573db32fda375d7fdebc437e37L55

Reply:
Right, I'll update this with that change.